src/main/java/functions/BreadthClosureFunction.kt

```diff
@@ -7,11 +7,13 @@ fun breadthClosureFunction(
 ): VisitorFunction {
     return { targetTreeNode ->
         var currentNode = targetTreeNode
-        val answer = mutableListOf<List<TargetTreeNode>>()
-        while (currentNode != null) {
-            answer.add(expression.invoke(currentNode).matches)
+        var currentAnswer = expression.invoke(currentNode)
+        val answer = mutableListOf<List<List<TargetTreeNode>>>(listOf(listOf()))
+        while (currentNode != null && currentAnswer.hasMatch) {
+            answer.add(currentAnswer.matches)
+            currentAnswer = expression.invoke(currentNode)
             currentNode = currentNode.moveToRightSibling()
         }
```

Comment:
I think there is a problem here, with the order in which currentNode and currentAnswer are calculated. I propose one solution that should fix this problem:

```suggestion
        var currentAnswer = expression.invoke(currentNode)
        val answer = mutableListOf<List<List<TargetTreeNode>>>(listOf(listOf()))
        while (currentAnswer.hasMatch) {
            answer.add(currentAnswer.matches)
            currentNode = currentNode!!.moveToRightSibling()
            if (currentNode == null) {
                currentAnswer = VisitorFunctionResponse()
            } else {
                currentAnswer = expression.invoke(currentNode)
            }
        }
```

Reply:
Yes, I think the same but the solution that I think is better is 
```suggestion
        var currentAnswer = expression.invoke(currentNode)
        val answer = mutableListOf<List<List<TargetTreeNode>>>(listOf(listOf()))
        while (currentNode != null && currentAnswer.hasMatch) {
            answer.add(currentAnswer.matches)
            currentNode = currentNode.moveToRightSibling()
            currentAnswer = expression.invoke(currentNode)
        }
```

Reply:
I did think of that, but I don't know what happens if `currentNode` is `null`, all functions are ready for it?